lib/collection/src/collection_manager/holders/proxy_segment.rs

```diff
@@ -863,6 +872,14 @@ impl SegmentEntry for ProxySegment {
             .read()
             .fill_query_context(query_context)
     }
+
+    fn is_proxy(&self) -> bool {
+        true
+    }
+
+    fn tmp_path(&self) -> PathBuf {
+        self.write_segment.get().read().data_path()
+    }
```

Comment:
The result of this function is not stored anywhere and is used immediately. Would it make sense to return a non-owning type here to avoid clones? E.g. something like this: `fn tmp_path(&self) -> MappedRwLockReadGuard<'_, Path<'_>>`. It would require a similar change for `fn data_path()`.

Reply:
Yeah, I think I'll make just that. Initially this was made for debug session, and I didn't care to refactor `SegmentEntry::data_path`, so I just rolled with `PathBuf`s everywhere.

But if we gonna merge this, I don't want to have allocating calls in the `upsert_points` path. I've tried to implement a check with `tracing` if particular span/event would be logged, and if not, avoid calling `SegmentEntry::id`... but neither of my approaches worked, so far. :/

Reply:
So, having a `tracing::info!("{}", something())` is not enough to not call the `something()` if it is disabled?

Reply:
So, I haven't strictly tested all of this, but:
- for message format arguments, it *might* be enough
- but I'm not using `SegmentEntry::id` as a format argument, I use it as a "field"
- and `tracing` allows using fields for *filtering*
- so, I assume, that fields *have to* always be evaluated
  - well, maybe except when the span/event is disabled by the plain log-level
  - but I explicitly use fields to filter tracing data in this PR
  - so I assume `segment.id = segment.id()` in the tracing stuff is always evaluated

To make it even worse:
- I tried a few approaches to pre-filter instrumentation
- e.g., query "will this span/event be enabled given it has this log-level and these fields"
- but in all cases `tracing` *always* reported instrumentation as *enabled*
  - even though instrumentation was *not* logged, i.e., it *was* filtered out
- so, I'd assume, that with how PR is structured right now, `tracing` evaluates *all* expressions first
  - and filtering applied later, once everything is already evaluated

Reply:
_Partially_ fixed this. `SegmentEntry::id` is cheap to call now, though, I assume `segment.id = %segment.id()` in instrumentation still allocates.

Reply:
Refactored my "persistent" `SegmentId` type and related methods. It should not require allocations now. At least, not in most "usual" cases.